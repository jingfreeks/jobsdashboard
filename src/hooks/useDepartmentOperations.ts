import { useMemo } from 'react';
import {
  useGetDepartmentsQuery,
  useAddDepartmentMutation,
  useUpdateDepartmentMutation,
  useDeleteDepartmentMutation,
  selectDepartmentsSorted,
  selectDepartmentById,
  type Department,
  type DepartmentFormData,
  type UpdateDepartmentData,
} from '@/features/department';

export const useDepartmentOperations = () => {
  const {
    data: departments = [],
    isLoading,
    error,
    refetch,
  } = useGetDepartmentsQuery();

  const [addDepartment, { isLoading: isAdding }] = useAddDepartmentMutation();
  const [updateDepartment, { isLoading: isUpdating }] = useUpdateDepartmentMutation();
  const [deleteDepartment, { isLoading: isDeleting }] = useDeleteDepartmentMutation();

  // Memoized sorted departments for performance
  const sortedDepartments = useMemo(() => selectDepartmentsSorted(departments), [departments]);

  // Memoized department map for quick lookups
  const departmentMap = useMemo(() => {
    const map = new Map<string, Department>();
    departments.forEach(dept => map.set(dept._id, dept));
    return map;
  }, [departments]);

  // Create a new department
  const createDepartment = async (departmentData: DepartmentFormData): Promise<Department | null> => {
    try {
      const result = await addDepartment(departmentData).unwrap();
      return result;
    } catch (error) {
      console.error('Failed to create department:', error);
      throw error;
    }
  };

  // Update an existing department
  const updateDepartmentById = async (departmentData: UpdateDepartmentData): Promise<Department | null> => {
    try {
      const result = await updateDepartment(departmentData).unwrap();
      return result;
    } catch (error) {
      console.error('Failed to update department:', error);
      throw error;
    }
  };

  // Delete a department
  const deleteDepartmentById = async (id: string): Promise<void> => {
    try {
      await deleteDepartment({ _id: id }).unwrap();
    } catch (error) {
      console.error('Failed to delete department:', error);
      throw error;
    }
  };

  // Get a department by ID
  const getDepartmentById = (id: string): Department | undefined => {
    return selectDepartmentById(departments, id);
  };

  // Search departments by name
  const searchDepartments = (searchTerm: string): Department[] => {
    if (!searchTerm.trim()) return sortedDepartments;
    const term = searchTerm.toLowerCase();
    return sortedDepartments.filter(dept => 
      dept.name.toLowerCase().includes(term)
    );
  };

  return {
    // Data
    departments: sortedDepartments,
    departmentMap,
    
    // Loading states
    isLoading,
    isAdding,
    isUpdating,
    isDeleting,
    
    // Error state
    error,
    
    // Operations
    createDepartment,
    updateDepartmentById,
    deleteDepartmentById,
    getDepartmentById,
    searchDepartments,
    
    // Utilities
    refetch,
  };
}; 