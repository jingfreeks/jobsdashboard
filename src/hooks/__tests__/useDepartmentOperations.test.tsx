import { describe, it, expect, vi, beforeEach } from 'vitest';
import { useDepartmentOperations } from '../useDepartmentOperations';

// Mock the department API
vi.mock('@/features/department', () => ({
  useGetDepartmentsQuery: vi.fn(),
  useAddDepartmentMutation: vi.fn(),
  useUpdateDepartmentMutation: vi.fn(),
  useDeleteDepartmentMutation: vi.fn(),
  selectDepartmentsSorted: vi.fn((data) => data?.sort((a, b) => a.name.localeCompare(b.name)) || []),
  selectDepartmentById: vi.fn((data, id) => data?.find((dept) => dept._id === id)),
}));

describe('useDepartmentOperations', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should be defined', () => {
    expect(useDepartmentOperations).toBeDefined();
  });

  it('should return a function', () => {
    expect(typeof useDepartmentOperations).toBe('function');
  });
}); 