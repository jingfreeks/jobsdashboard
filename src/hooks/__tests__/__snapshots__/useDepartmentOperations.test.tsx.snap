// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDepartmentOperations (Vitest) > should return sorted departments and departmentMap 1`] = `
{
  "createDepartment": [Function],
  "deleteDepartmentById": [Function],
  "departmentMap": Map {
    "1" => {
      "_id": "1",
      "name": "HR",
    },
    "2" => {
      "_id": "2",
      "name": "Engineering",
    },
  },
  "departments": [
    {
      "_id": "1",
      "name": "HR",
    },
    {
      "_id": "2",
      "name": "Engineering",
    },
  ],
  "error": null,
  "getDepartmentById": [Function],
  "isAdding": false,
  "isDeleting": false,
  "isLoading": false,
  "isUpdating": false,
  "refetch": [MockFunction spy],
  "searchDepartments": [Function],
  "updateDepartmentById": [Function],
}
`;
